### DISCUSSION TITLE
Proposal to Develop an NTRIP Server Extension for Tasmota

### DISCUSSION THREAD

**Discussion by catalinsanda** [arendst/Tasmota#21688] [Ideas]
Hi folks,

I've got an idea that I'd love to get your thoughts on. I'm thinking about developing an extension to Tasmota that would allow it to act as an NTRIP server for GNSS boards. I’m starting with a Beitian board using the Unicore UM980 chip since that’s what I have on hand, but the goal is to make it work with any GNSS board.

The main idea is to enable Tasmota to support GNSS functionality and upload RTK correction data to networks like RTK2GO. I’ve noticed there are standalone NTRIP server solutions out there, like the https://github.com/slgrobotics/Esp32_RTK_BaseStation project, but they often require custom firmware. By integrating this into Tasmota, we could make it much easier to configure through the Tasmota UI.

I'm also thinking ahead about adding support for a rover and using Tasmota’s LoRa capabilities to link the rover with the base. This could open up a lot of new applications in areas like surveying, agriculture, and autonomous vehicles.

So, I wanted to ask:
* Does this idea fit with what Tasmota is all about?
* Are there any technical issues or existing plans I should know about?
* Would you be open to accepting contributions in this direction?

If there’s interest and support, I’m ready to start working on a proof of concept for the UM980 chip. Any guidance or suggestions would be really appreciated, and I’m excited to collaborate with anyone else who might be interested.

Thanks for taking the time to read this. Looking forward to your thoughts!

Best,
Catalin

**Comment by joba-1** [comment]
A rather exotics use case I happened to start caring about recently :)
Have you heard about the Quectel chips lc29h? (the only ones I looked at so far…)

Regarding your question: I‘m a bit concerned about timing requirements for rtk because that is not a strength of tasmota. But I may be wrong about this.

**Comment by catalinsanda** [comment]
Hi everyone!

About a year ago, I posted here asking for thoughts on developing a Tasmota extension for GNSS/NTRIP functionality. Well, I'm excited to share that I've successfully implemented it!

## What I've Built

I've created a comprehensive GNSS and NTRIP driver system for Tasmota that includes:
- Full GNSS support with multi-constellation tracking
- Dual NTRIP server functionality (connect to multiple NTRIP caster services)
- Local NTRIP caster capability (serve corrections to rovers)
- Real-time RTCM processing and HTTP data streaming
- Web-based configuration and monitoring

## Where to Find It

- **Code Repository:** https://github.com/catalinsanda/Tasmota (branch: `catalinsanda/rtk-server`)
- **Complete Project Guide:** https://hackaday.io/project/203323-gnss-base-station-with-ntrip-server-for-tasmota

The Hackaday project includes everything you need: hardware assembly instructions, firmware building/flashing guide, configuration steps, and even how to achieve millimeter-precision coordinates using PPP services.

## Integration Status

I attempted to contribute this back to the main Tasmota repository via PR #23028 (https://github.com/arendst/Tasmota/pull/23028), but it was rejected because the implementation requires a secondary HTTP server for the NTRIP caster functionality. While I understand the architectural concerns, the driver works great as a community fork.

## Hardware Tested

- OLIMEX ESP32-POE2 + UM980-based GNSS modules (WTRTK-980)
- Custom PCB design available for clean integration
- PoE support makes remote base station deployment easy

The system is working excellently in production - I'm currently contributing corrections to RTK2GO and Centipede-RTK and also serving local rovers with centimeter-level accuracy.

**If you're interested in precision GNSS applications, I'd love for you to try it out!** The project makes it incredibly easy to set up professional-grade RTK base stations using familiar Tasmota infrastructure.

Best,
Catalin

**Reply by pkkrusty** [reply]
Excellent! I've been very happy with a basic setup using a Quectel LC29HEA and a raspberry pi zero 2 W as a base station serving NTRIP data to rtk2go. That is a $100 solution. I'll give your fork a try!

**Reply by catalinsanda** [reply]
Thanks! All my testing was done with UM980-based boards, and the setup described on Hackaday is tailored for these modules. However, the drivers themselves aren’t specific to the UM980, so there’s a very good chance it will work just fine with your LC29HEA setup as well. I'd be very interested to hear how it goes if you give it a try!
